File: runtime-elixir/server/smithy_error_map.ex
defmodule SmithyErrorMap do
  @moduledoc """
  Maps Smithy error terms to `{http_status, message}` pairs.

  This is the generic fallback module used when a service has no modelled errors.
  For services that define error shapes with `@httpError`, the code generator
  produces a per-service override that prepends specific clauses for each modelled
  error before falling through to these defaults.

  Mirrors `smithy_error_map.erl` from the Erlang runtime.
  """

  @doc """
  Map a Smithy error term to `{http_status_code, message}`.

  ## Error terms

  | Term                      | Status | Notes                        |
  |---------------------------|--------|------------------------------|
  | `{:not_found, msg}`       | 404    |                              |
  | `{:conflict, msg}`        | 409    |                              |
  | `{:validation, msg}`      | 400    |                              |
  | `{:internal, msg}`        | 500    |                              |
  | `{:unauthorized, msg}`    | 401    |                              |
  | `{:forbidden, msg}`       | 403    |                              |
  | `:not_implemented`        | 501    |                              |
  | anything else             | 500    | Generic internal server error |
  """
  @spec to_http(term()) :: {pos_integer(), String.t()}
  def to_http({:not_found, msg}),    do: {404, msg}
  def to_http({:conflict, msg}),     do: {409, msg}
  def to_http({:validation, msg}),   do: {400, msg}
  def to_http({:internal, msg}),     do: {500, msg}
  def to_http({:unauthorized, msg}), do: {401, msg}
  def to_http({:forbidden, msg}),    do: {403, msg}
  def to_http(:not_implemented),     do: {501, "Not implemented"}
  def to_http(_),                    do: {500, "Internal server error"}
end

File: runtime-erlang/server/smithy_error_map.erl
-module(smithy_error_map).

%% Maps Smithy error atoms/tuples to HTTP status codes and message binaries.
%%
%% This is the generic fallback module used when a service has no modeled errors.
%% For services that do define error shapes with @httpError, the code generator
%% produces a per-service override of this module that prepends specific clauses
%% for each modeled error before falling through to these defaults.

-export([to_http/1]).

%%%===================================================================
%%% API Functions
%%%===================================================================

%% @doc Map a Smithy error term to {HttpStatusCode, MessageBinary}.
%%
%% Supported error terms:
%%   {not_found, Msg}    -> 404
%%   {conflict, Msg}     -> 409
%%   {validation, Msg}   -> 400
%%   {internal, Msg}     -> 500
%%   {unauthorized, Msg} -> 401
%%   {forbidden, Msg}    -> 403
%%   not_implemented     -> 501
%%   _                   -> 500 (catch-all)
-spec to_http(Err :: term()) -> {pos_integer(), binary()}.
to_http({not_found, Msg})    -> {404, Msg};
to_http({conflict, Msg})     -> {409, Msg};
to_http({validation, Msg})   -> {400, Msg};
to_http({internal, Msg})     -> {500, Msg};
to_http({unauthorized, Msg}) -> {401, Msg};
to_http({forbidden, Msg})    -> {403, Msg};
to_http(not_implemented)     -> {501, <<"Not implemented">>};
to_http(_)                   -> {500, <<"Internal server error">>}.
